operator: Support upgrade of plugins

ozhuraki opened the pull request:
Support upgrade of plugins' images. e.g. upgrading them to the current official release of the operator.

This is done only once at the start of the operator (the new image version is stored into the CR).

This feature is optional and can be controlled through the command line parametes.

Closes #702

mythi commented on pkg/controllers/dsa/controller.go:
```diff
@@ -63,12 +65,26 @@ type controller struct {
 	controllers.DefaultServiceAccountFactory
 	scheme *runtime.Scheme
 	ns     string
+	log    logr.Logger
```
`Logger` is available from context so I belive this isn't needed.

mythi commented on pkg/controllers/dsa/controller.go:
```diff
@@ -63,12 +65,26 @@ type controller struct {
+	dp := obj.(*devicepluginv1.DsaDevicePlugin)
+
+	if img := controllers.UpgradeToImage(dp.Spec.Image); len(img) > 0 {
+		dp.Spec.Image = img
```
dsa, sgx, fpga, and gpu have `initImage`s to update too.

mythi commented on pkg/controllers/reconciler.go:
```diff
@@ -189,26 +258,33 @@ func (r *reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 
 	ds := &childDaemonSets.Items[0]
 
+	if UpgradePlugins {
```
would it make sense to set `func()` based on `UpgradePlugins`. This would save complexity.

mythi reviewed (changes requested):
the approach looks good to me. please check `initImage` upgrade and linter checks.

mythi commented on pkg/controllers/reconciler.go:
```diff
@@ -159,6 +187,47 @@ func (r *reconciler) createObjects(ctx context.Context,
+	out := ""
+
+	if len(UpgradeToVersion) > 0 {
+		out = imageGetName(image) + UpgradeToVersion
```
The syntax for `UpgradeToVersion` is not checked. It should be done already during startup that we get to notify user as soon as possible.

ozhuraki commented:
@mythi
Thanks for the input, updated, please take a look. Changes:
- Re-worked to use Logger from context
- Added handling of initimages for DSA, FPGA, GPU, SGX
- Re-worked invocation of sync.Once.Do() in Reconcile()
- Added a syntax check for UpgradeToVersion

rojkov commented on pkg/controllers/reconciler.go:
```diff
@@ -17,28 +17,42 @@ package controllers
 	bKeeper.pluginCounter = make(map[string]int)
 }
 
+type Once struct {
```
nit: I'd rather hide this type and the variables like `UpgradePlugins`, `UpgradeToVersion` from public interface. BTW, not sure this type helps readability. Otherwise looks very good.

ozhuraki replied:
@rojkov @mythi 
> I'd rather hide this type [...] BTW, not sure this type helps readability

Thanks for the input and fully agree. The problem here is linter (which ATM simply counts "if"s, and Reconcile() is already at the limit), so this trickery is just tame lint.

I had a few experiments to reduce cyclo whithout harming readability of Reconcile(), but wasn't too pleased with results (brainstorming it and will return with some solution after this PR).

Let me check if these simplest fixes would do.

mythi commented on pkg/controllers/reconciler.go:
```diff
@@ -159,6 +174,61 @@ func (r *reconciler) createObjects(ctx context.Context,
+
+	if UpgradePlugins {
+		o.once.Do(func() {
+			if r.controller.Upgrade(ctx, devicePlugin) {
```
would this work?
```suggestion
			if UpgradePlugins && r.controller.Upgrade(ctx, devicePlugin) {
```

mythi replied:
(assuming it's still valid after @rojkov's comments)

ozhuraki commented:
@mythi @rojkov
Updated, please take a look, changes:
- Once -> wrapOnce. I tried to eliminate it, but can't come up with the inline logic with one "if" (we need an err from Update())
- Simplified the logic in Upgrade()

rojkov commented on pkg/controllers/reconciler.go:
```diff
@@ -159,6 +174,59 @@ func (r *reconciler) createObjects(ctx context.Context,
+	return out
+}
+
+func (o *wrapOnce) Upgrade(ctx context.Context, r *reconciler, devicePlugin client.Object) {
+	log := log.FromContext(ctx)
+
+	o.once.Do(func() {
+		if UpgradePlugins && r.controller.Upgrade(ctx, devicePlugin) {
+			if err := r.Update(ctx, devicePlugin); err != nil {
+				log.Error(err, "unable to update devicePlugin")
+			}
+		}
+	})
+}
```
If this function depends on global state already (meaning `UpgradePlugins`) then using the global `once` directly won't make the situation worse, I guess. Like in
```suggestion
func upgradeOnce(ctx context.Context, r *reconciler, devicePlugin client.Object) {
	log := log.FromContext(ctx)

	once.Do(func() {
		if UpgradePlugins && r.controller.Upgrade(ctx, devicePlugin) {
			if err := r.Update(ctx, devicePlugin); err != nil {
				log.Error(err, "unable to update devicePlugin")
			}
		}
	})
}
```

ozhuraki replied:
@rojkov 
Thanks! Dropped wrapOnce.

rojkov commented:
Looks like the coverage went down a bit. Can we cover the new code?

mythi commented on pkg/controllers/reconciler.go:
```diff
@@ -189,8 +253,16 @@ func (r *reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 
 	ds := &childDaemonSets.Items[0]
 
+	if UpgradePlugins {
```
if upgrade is meant to be called only one, why check this always?

mythi replied:
@ozhuraki once the upgrade is done, the check is no longer relevant. Check the suggestion above

bart0sh commented on cmd/operator/main.go:
```diff
@@ -101,20 +103,29 @@ func main() {
 		"Enable leader election for controller manager. "+
 			"Enabling this will ensure there is only one active controller manager.")
+	flag.BoolVar(&controllers.UpgradePlugins, "upgrade", true, fmt.Sprintf("Auto-upgrade plugins' images to the current release %s", controllers.ImageMinVersion))
+	flag.StringVar(&controllers.UpgradeToVersion, "upgrade-to-version", controllers.ImageMinVersion, "Upgrade / downgrade to official version, e.g. X.Y.Z, 0.23.0")
```
Have I understood correctly that we're changing 2 variables in the controllers module this way?
Can we use local variables here and pass them along as parameters instead?

mythi commented on pkg/controllers/reconciler.go:
```diff
@@ -39,6 +39,7 @@ import (
 var (
 	UpgradePlugins   bool
 	UpgradeToVersion string
+	EnvtestMode      bool
```
> Apparently controller-runtime caches read operations,
> injecting a daemonset for upgrade envtest is problematic without it.

how this works in your manual testing?

ozhuraki replied:
@mythi 
> how this works in your manual testing?

Once the controller is instantiated, it does some indexing to cache read operations. In manual tests it is able to find the previous daemonset.

With the injection of the daemonset under envtest, apparently not. I am not entirely sure what's the subtle difference, either the controller should be deleted and recreated before each upgrade envtest (didn't succeed with this) or some subtle detail in indexing.

mythi replied:
@ozhuraki would this type of test work better under the e2e test framework then? I don't like the fact that we need to modify the controller's behavior depending on whether it's under test or not.

hj-johannes-lee commented on test/envtest/dlbdeviceplugin_controller_test.go:
```diff
@@ -102,4 +105,49 @@ var _ = Describe("DlbDevicePlugin Controller", func() {
+
+		Expect(client.Delete(ctx, dp) == nil)
+
+		time.Sleep(time.Second)
```
Hello.! Can't this be replaced with const "interval"?

hj-johannes-lee commented on test/envtest/dlbdeviceplugin_controller_test.go:
```diff
@@ -102,4 +105,49 @@ var _ = Describe("DlbDevicePlugin Controller", func() {
+		ds0.ObjectMeta.Namespace = ns
+		ds0.Spec.Template.Spec.Containers[0].Image = image0
+
+		Expect(client.Create(ctx, ds0) == nil)
```
I don't know about ginkgo that much, would it look nicer to have 'By("blah blah")'?

ozhuraki commented:
@mythi @rojkov
Changes, please take a look:
- Re-worked envtests to drop the controller's test toggle
- Added a reconciler test with a few simple checks
- Dropped UpgradeToVersion command line parameter

bart0sh commented on pkg/controllers/fpga/controller.go:
```diff
@@ -65,6 +66,27 @@ func (c *controller) CreateEmptyObject() client.Object {
 	return &devicepluginv1.FpgaDevicePlugin{}
 }
 
+func (c *controller) Upgrade(ctx context.Context, obj client.Object) (upgrade bool) {
```
This function seem to be quite generic. Can we avoid code duplication if we move the code into the function that accepts an interface with a couple of methods like GetImage and GetInitImage?

ozhuraki replied:
@bart0sh 
In principle this could be done, but we would need 4 methods in each device controller, i.e. GetImage, GetInitImage, SetImage, SetInitImage. Should we go for it?

mythi commented on pkg/controllers/reconciler_test.go:
```diff
@@ -0,0 +1,37 @@
+)
+
+func TestUpgrade(t *testing.T) {
+	in := "intel/intel-dsa-plugin:0.22.0"
```
also test with, e.g., `:latest` and without version at all?

mythi commented on test/envtest/gpudeviceplugin_controller_test.go:
```diff
@@ -20,20 +20,25 @@ import (
 	"k8s.io/apimachinery/pkg/types"
 
 	devicepluginv1 "github.com/intel/intel-device-plugins-for-kubernetes/pkg/apis/deviceplugin/v1"
+	"github.com/intel/intel-device-plugins-for-kubernetes/pkg/controllers"
```
nit: can you check these import paths are consistent in `*deviceplugin_controller_test.go` files?

bart0sh commented on pkg/controllers/gpu/controller.go:
```diff
@@ -70,6 +71,27 @@ func (c *controller) CreateEmptyObject() client.Object {
+
+	return upgrade
+}
+
```
I was thinking about something as simple as this:
```
func (c *controller) UpgradeImages(ctx context.Context, image, initImage string) (string, string, bool) {
        log := log.FromContext(ctx)
        if img := controllers.UpgradeToImage(image); len(image) > 0 {
		upgrade = true
		image = img

		log.Info("Upgrade to image", img)
	}

	if img = controllers.UpgradeToImage(initImage); len(img) > 0 {
		upgrade = true
		initImage = img

		log.Info("Upgrade to initimage", initImage)
	}

        return image, initImage, upgrade
}

func (c *controller) Upgrade(ctx context.Context, obj client.Object) (upgrade bool) {
	dp := obj.(*devicepluginv1.GpuDevicePlugin)
         dp.Spec.Image, dp.Spec.InitImage, upgrade := c.UpgradeImages(ctx, dp.Spec.Image, dp.Spec.InitImage)
         return upgrade
}
```

ozhuraki commented:
@mythi @bart0sh
Changes, please take a look:
- De-duplicated controllers' code (@bart0sh thanks for the idea)
- Updated the test with latest and no version

mythi commented on pkg/controllers/reconciler.go:
```diff
@@ -17,24 +17,32 @@ package controllers
 
 var (
-	bKeeper = &bookKeeper{}
+	UpgradePlugins bool
```
If we set a default value here, the test cases do not need to import and set the value

ozhuraki replied:
@mythi 
In principle, not a bad idea, I just tried, but there's a problem to create initial 0.22.0 CRs in envtests, they are upgraded immediately. I.e. in one way or another this should be controlled in envtests.

bart0sh commented on cmd/operator/main.go:
```diff
@@ -101,13 +102,14 @@ func main() {
 	flag.BoolVar(&enableLeaderElection, "enable-leader-election", false,
 		"Enable leader election for controller manager. "+
 			"Enabling this will ensure there is only one active controller manager.")
+	flag.BoolVar(&controllers.UpgradePlugins, "upgrade", true, fmt.Sprintf("Auto-upgrade plugins' images to the current release %s", controllers.ImageMinVersion))
```
I'm still thinking that this is not very good to set module variable this way. Can we add Upgradable call to the controller interface (see example below)? This is more wordy, but gives more flexibility in my opinion. For example, with this approach it's possible to configure only one or subset of plugins to be upgradable. Not sure it makes practical sense though, but it at least makes sense to discuss it. @mythi @rojkov @ozhuraki  WDYT?
```
diff --git a/cmd/operator/main.go b/cmd/operator/main.go
index c25637c6..66f25be6 100644
--- a/cmd/operator/main.go
+++ b/cmd/operator/main.go
@@ -57,7 +57,7 @@ func init() {
        // +kubebuilder:scaffold:scheme
 }
 
-type devicePluginControllerAndWebhook map[string](func(ctrl.Manager, string, bool) error)
+type devicePluginControllerAndWebhook map[string](func(ctrl.Manager, string, bool, bool) error)
 
 type flagList []string
 
@@ -100,6 +100,7 @@ func main() {
                devicePluginNamespace string
                enableLeaderElection  bool
                pm                    *patcher.Manager
+               upgradePlugins        bool
        )
 
        ctrl.SetLogger(klogr.NewWithOptions(klogr.WithFormat(klogr.FormatKlog)))
@@ -109,7 +110,7 @@ func main() {
        flag.BoolVar(&enableLeaderElection, "enable-leader-election", false,
                "Enable leader election for controller manager. "+
                        "Enabling this will ensure there is only one active controller manager.")
-       flag.BoolVar(&controllers.UpgradePlugins, "upgrade", true, fmt.Sprintf("Auto-upgrade plugins' images to the current release %s", controllers.ImageMinVersion))
+       flag.BoolVar(&upgradePlugins, "upgrade", true, fmt.Sprintf("Auto-upgrade plugins' images to the current release %s", controllers.ImageMinVersion))
        flag.Var(&devices, "devices", "Device(s) to set up.")
        flag.Parse()
 
@@ -147,7 +148,7 @@ func main() {
        withWebhook := true
 
        for _, device := range devices {
-               if err = setupControllerAndWebhook[device](mgr, ns, withWebhook); err != nil {
+               if err = setupControllerAndWebhook[device](mgr, ns, withWebhook, upgradePlugins); err != nil {
                        setupLog.Error(err, "unable to initialize controller", "controller", device)
                        os.Exit(1)
                }
diff --git a/pkg/controllers/dlb/controller.go b/pkg/controllers/dlb/controller.go
index e027a710..3244a27c 100644
--- a/pkg/controllers/dlb/controller.go
+++ b/pkg/controllers/dlb/controller.go
@@ -42,8 +42,8 @@ var defaultNodeSelector map[string]string = deployments.DLBPluginDaemonSet().Spe
 // +kubebuilder:rbac:groups=deviceplugin.intel.com,resources=dlbdeviceplugins/finalizers,verbs=update
 
 // SetupReconciler creates a new reconciler for DlbDevicePlugin objects.
-func SetupReconciler(mgr ctrl.Manager, namespace string, withWebhook bool) error {
-       c := &controller{scheme: mgr.GetScheme(), ns: namespace}
+func SetupReconciler(mgr ctrl.Manager, namespace string, withWebhook, upgradable bool) error {
+       c := &controller{scheme: mgr.GetScheme(), ns: namespace, upgradable: upgradable}
        if err := controllers.SetupWithManager(mgr, c, devicepluginv1.GroupVersion.String(), "DlbDevicePlugin", ownerKey); err != nil {
                return err
        }
@@ -57,8 +57,9 @@ func SetupReconciler(mgr ctrl.Manager, namespace string, withWebhook bool) error
 
 type controller struct {
        controllers.DefaultServiceAccountFactory
-       scheme *runtime.Scheme
-       ns     string
+       scheme     *runtime.Scheme
+       ns         string
+       upgradable bool
 }
 
 func (c *controller) CreateEmptyObject() client.Object {
@@ -153,6 +154,10 @@ func (c *controller) UpdateStatus(rawObj client.Object, ds *apps.DaemonSet, node
        return updated, nil
 }
 
+func (c *controller) Upgradable() bool {
+       return c.upgradable
+}
+
 func getPodArgs(gdp *devicepluginv1.DlbDevicePlugin) []string {
        args := make([]string, 0, 4)
        args = append(args, "-v", strconv.Itoa(gdp.Spec.LogLevel))

diff --git a/pkg/controllers/reconciler.go b/pkg/controllers/reconciler.go
index f0e5863c..28e25d81 100644
--- a/pkg/controllers/reconciler.go
+++ b/pkg/controllers/reconciler.go
@@ -36,10 +36,7 @@ import (
        "sigs.k8s.io/controller-runtime/pkg/log"
 )
 
-var (
-       UpgradePlugins bool
-       bKeeper        = &bookKeeper{}
-)
+var bKeeper = &bookKeeper{}
 
 const ImageMinVersion = "0.23.0"
 
@@ -106,6 +103,7 @@ type DevicePluginController interface {
        UpdateDaemonSet(client.Object, *apps.DaemonSet) (updated bool)
        UpdateStatus(client.Object, *apps.DaemonSet, []string) (updated bool, err error)
        Upgrade(ctx context.Context, obj client.Object) (upgrade bool)
+       Upgradable() bool
 }
 
 type reconciler struct {
@@ -226,7 +224,7 @@ func UpgradeImages(image *string, initimage *string) (upgrade bool) {
 
 func upgradeOnce(ctx context.Context, r *reconciler, devicePlugin client.Object) {
        r.once.Do(func() {
-               if UpgradePlugins && r.controller.Upgrade(ctx, devicePlugin) {
+               if r.controller.Upgradable() && r.controller.Upgrade(ctx, devicePlugin) {
                        if err := r.Update(ctx, devicePlugin); err != nil {
                                log := log.FromContext(ctx)
                                log.Error(err, "unable to update devicePlugin")

```

rojkov replied:
I second it. At least it may make testing more convenient and less prone to races. 

Was going to comment on that, but got distracted.

ozhuraki replied:
@rojkov @mythi @bart0sh 
It boils down to whether the upgrade should be per device or global. At the moment it's fairly simple - global and minimally controllable (on/off, if there's a need).

More fine-grained control:
- Needed in practice?
- Would be simple in understanding and usage?

Not sure. Thoughts?

rojkov replied:
Being global and RW-accessible from anywhere is the problem. `controller.UpgradePlugins` adds global state and introduces a public interface to it which is supposed to be supported *forever*. And we just don't know currently what kind of problems we are going to have down the road. Consider multiple test cases running in parallel in one test process and updating/reading the state. Obviously the cases wouldn't be isolated enough from each other.

rojkov commented on test/envtest/suite_test.go:
```diff
@@ -59,57 +62,66 @@ func TestAPIs(t *testing.T) {
+
+	Expect(devicepluginv1.AddToScheme(scheme.Scheme) == nil)
+
+	// +kubebuilder:scaffold:scheme
```
The code in this file used to be auto-generated by kubebuilder scaffolding. I guess this tag can be dropped now.

rojkov commented on pkg/controllers/reconciler_test.go:
```diff
@@ -0,0 +1,78 @@
+		t.Errorf("expectedImage: %s, received: %s", expectedImage, image)
+		t.Errorf("expectedInitimage: %s, received: %s", expectedInitimage, initimage)
+	}
+}
```
Nit: this looks like a good fit for a table-driven test (e.g. like in cmd/fpga_plugin/fpga_plugin_test.go).

rojkov reviewed (commented):
Thanks! Looks great. Just added a couple of nits.

ozhuraki commented:
@rojkov @mythi @bart0sh
Changes, please take a look:
- UpgradePlugins -> upgradable, per controller now
- Dropped kubebuilder's scaffold:scheme in test/envtest/suite_test.go
- Converted pkg/controllers/reconciler_test.go to table

bart0sh commented on cmd/operator/README.md:
```diff
@@ -144,6 +145,12 @@ In this case, create a new kustomization with the necessary resources
+
+The operator supports auto-upgrade of plugins' images. e.g. upgrading them to the current official release of the operator. This is done only once at the start of the operator (the new image / version is stored into the CR).
+
+This feature is optional and can be controlled through the command line parametes.
```
typo: parametes -> parameters

bart0sh commented on pkg/controllers/reconciler.go:
```diff
@@ -159,6 +167,72 @@ func (r *reconciler) createObjects(ctx context.Context,
+	minVer, e2 := version.ParseSemantic(ImageMinVersion)
+
+	if e1 == nil && e2 == nil && ver.LessThan(minVer) {
+		out = "intel/" + imageGetName(image) + ":" + ImageMinVersion
```
nit: if imageGetName returns empty string we'll get 'intel/:<version>' as a result.

bart0sh commented on pkg/controllers/reconciler.go:
```diff
@@ -159,6 +167,72 @@ func (r *reconciler) createObjects(ctx context.Context,
+			upgrade = true
+		}
+	}
+
```
nit: would it make sense to convert this function into a loop with 2 iterations and get rid of upgradeToImage function?

mythi commented on cmd/operator/main.go:
```diff
@@ -145,7 +148,7 @@ func main() {
 	withWebhook := true
```
it's false in envtests because we don't register webhooks there.

mythi replied:
see 37618d4f850f63ae0503b0b313242540a9bb269d

bart0sh replied:
I understand, just wondering if we need a variable for it here.

ozhuraki commented:
@bart0sh
Changes:
- Fixed a typo in the documentation
- Added a check for an empty image length

bart0sh commented on pkg/apis/deviceplugin/v1/dlbdeviceplugin_webhook.go:
```diff
@@ -33,7 +33,7 @@ var (
 	dlbdevicepluginlog = logf.Log.WithName("dlbdeviceplugin-resource")
 
-	dlbMinVersion = version.MustParseSemantic(imageMinVersion)
+	dlbMinVersion = version.MustParseSemantic(controllers.ImageMinVersion)
```
would it make sense to call version.MustParseSemantic once in controllers module?

ozhuraki commented:
@bart0sh
Changes:
- Make withWebhook controllable from the command line

mythi commented:
> Make withWebhook controllable from the command line

we don't want that

rojkov commented on test/envtest/sgxdeviceplugin_controller_test.go:
```diff
@@ -88,4 +93,41 @@ var _ = Describe("SgxDevicePlugin Controller", func() {
+
+		time.Sleep(time.Second)
+
+		down()
+
+		up("sgx", true)
```
Why this restart is needed? Can we have the operator running always, then create a plugin with an outdated image and check it again if it has the image field updated?

rojkov commented on test/envtest/sgxdeviceplugin_controller_test.go:
```diff
@@ -20,20 +20,25 @@ import (
-	AfterEach(func() {
-		time.Sleep(time.Second * 2)
+	BeforeEach(func() {
+		up("sgx", false)
```
Not sure we want to test all controllers isolated from each other in integration tests. The expected setup is that users run the operator with multiple controllers enabled.

ozhuraki commented:
@rojkov 
> What prevents us from upgrading the CR upon the first call?

Thanks, very good idea, I will update accordingly.

ozhuraki commented:
@mythi 
> Just remember the use case: existing CRs must be updated automatically when the operator starts

Thanks, important detail indeed, will re-check this!

ozhuraki commented:
@rojkov @mythi @bartosh

Changes for a preview, take a look at the last commit (will be squashed).

> - What prevents us from upgrading the CR upon the first call?
> - Only existing CRs must be updated automatically when the operator starts

These are mutually exclusive, so the restart in envtests is needed after all.

Also the control of the "upgradable" state by reconciler is required. Found and fixed a bug where the new CRs were upgraded too (the Reconcile() is called multiple times on CR creation).

The setters/getters for "upgradalbe" look redundant, shall we simply move the "upgradable" to the reconciler struct?

bart0sh commented on pkg/controllers/dlb/controller.go:
```diff
@@ -148,6 +154,14 @@ func (c *controller) UpdateStatus(rawObj client.Object, ds *apps.DaemonSet, node
+func (c *controller) SetUpgradable(upgradable bool) {
+	c.upgradable = upgradable
+}
+
```
Would it make sense to move upgradable attribute and above 2 methods to the separate struct and then inherit controller struct from it? This way we could avoid duplicating these methods for every controller.

bart0sh commented:
> The setters/getters for "upgradalbe" look redundant, shall we simply move the "upgradable" to the reconciler struct?

That would make sense to do unless we want to be able to upgrade only certain plugins. It doesn't seem to be the case, so +1 for moving to reconciler.

ozhuraki commented:
> That would make sense to do unless we want to be able to upgrade only certain plugins. It doesn't seem to be the case, so +1 for moving to reconciler.

This would actually be per plugin too, so no loss here.

> IMO, this upgradable is becoming too dominant in this change especially when the primary purpose of it is to get the tests working. If we cannot get envtest to reflect the real-world scenario, would the e2e fw be better?

Turns the "upgradable" is needed for "only existing CRs must be updated automatically when the operator starts".

> create the CR and start the controller afterwards?

Without a daemonset, this won't be any different from creating a CR with a controller already started (from the Reconcile()'s point of view).

We would run into the same in e2e, envtest is identical from this point.

2 runs of the controller to test the upgrade are needed and reflect the real world, unless I am missing anything?

rojkov commented:
I'm trying to figure what's going on in the real world. Am I correct assuming that an upgrade happens at the moment a new version of the operator starts? Like

* Precondition:  an old operator is running, a plugin's CR is in etcd.
* Step 1. The old operator gets stopped, its pod is deleted. The plugin's CR is still in etcd unchanged.
* Step 2. A new version of operator is added in the cluster with the minVersion greater than the CR's.
* Step 3. `Reconcile()` of the new operator gets called for the current (non-updated) CR.
* Step 4. The new operator detects the "oldness" and updates the CR and the DaemonSet linked to the CR in etcd.
* Step 5. `Reconcile()` gets called again for the updated versions of the CR and/or the DaemonSet. The operator does nothing this time.

If this scenario is correct I'd expect that a test could have an outdated CR (and DaemonSet probably) in etcd as a precondition before starting the operator.

If the scenario is wrong and we miss the opportunity to update upon the first call then, perhaps, it's better to keep `Reconcile()` idempotent and not to use `sync.Once`.

mythi commented:
> Am I correct assuming that an upgrade happens at the moment a new version of the operator starts? Like

@rojkov correct, the steps 1.-5. you summarized is what happens

ozhuraki commented:
@rojkov @mythi @bart0sh

> If this scenario is correct I'd expect that a test could have an outdated CR (and DaemonSet probably) in etcd as a precondition before starting the operator.

This scenario is correct, i.e. both an outdated CR and a daemonset are a precondition for upgrade.

We already tried to create such precondition in the tests by manually injecting a daemonset. On the operator side this required bypassing indexing for discovering a daemonset (we dropped this approach). There might be a subtle bug in how the indexing is setup / used on the operator side.

At present the first operator run is used primarily to inject a daemonset.

Otherwise we don't miss an opportunity to upgrade upon the first call.

At present, the Reconcile() isn't idempotent, due to multiple calls of the Reconcile() at Step 3, also on CR creation (there might be a subtle bug there too).

In order to handle the case "only existing CRs must be updated automatically when the operator starts", the "upgradable" would be reqiured in real life too (i.e. outside of envtests).

In principle, having just the "upgradable" alone, would be sufficient (i.e. we could drop sync.Once).

From the upgade point of view, so far I don't see how the Reconcile() could be made idempotent (nor it was / is an idempotent), correct me if I am wrong.

ozhuraki commented:
@rojkov @mythi @bartosh
Changes:
- Fixed a bug with upgrade on a CR creation. Now only existing CRs are upgraded (i.e. with a daemonset)
- Eliminated upgradable
- Removed hardcoding of version / previous version in upgrade tests
- Fixed a bug where initcontainer image wasn't updated in SGX controller

rojkov reviewed (commented):
> We already tried to create such precondition in the tests by manually injecting a daemonset. On the operator side this required bypassing indexing for discovering a daemonset (we dropped this approach). There might be a subtle bug in how the indexing is setup / used on the operator side.
> At present the first operator run is used primarily to inject a daemonset.

Is it possible to inject a daemonset and to index it in the test setup function without touching the operator?

> In order to handle the case "only existing CRs must be updated automatically when the operator starts", the "upgradable" would be reqiured in real life too (i.e. outside of envtests).

I guess I miss the context here. How non-existing CRs can be upgraded at all?

> At present, the Reconcile() isn't idempotent, due to multiple calls of the Reconcile() at Step 3, also on CR creation (there might be a subtle bug there too).
> From the upgade point of view, so far I don't see how the Reconcile() could be made idempotent (nor it was / is an idempotent), correct me if I am wrong.

You're right. I should have mentioned that for `Reconcile()` I include the state of etcd in the function's input and output (but not `reconciler`'s state).

ozhuraki commented:
@rojkov @mythi @bartosh
> Is it possible to inject a daemonset and to index it in the test setup function without touching the operator?

Without touching the operator not sure, I didn't spot an explicit index information in the daemonset itself. Ideas?

> I guess I miss the context here. How non-existing CRs can be upgraded at all?

There's a mixup here. @mythi meant that when we create a CR the first time (i.e. with the operator started and no plugin deployed, it shouldn't be upgraded), @mythi, right?

> The operator starts for the first time, notices in Reconcile() that the CR is missing its DS, disables upgrade, creates DS -> the CR is never upgraded.

From the Reconcile()'s point, this is a case when the CR is created initially, i.e. taking into the account above, this is as intended.

> upgradable never gets false. Better drop it to reduce the patch.

It's actually used: https://github.com/ozhuraki/intel-device-plugins-for-kubernetes/blob/operator-upgrade/cmd/operator/main.go#L151

mythi commented on cmd/operator/main.go:
```diff
@@ -99,15 +100,17 @@ func main() {
 	flag.BoolVar(&enableLeaderElection, "enable-leader-election", false,
 		"Enable leader election for controller manager. "+
 			"Enabling this will ensure there is only one active controller manager.")
+	flag.BoolVar(&upgradePlugins, "upgrade", true, fmt.Sprintf("Auto-upgrade plugins' images to the current release %s", controllers.ImageMinVersion.String()))
```
how about:
```suggestion
	flag.BoolVar(&upgradePlugins, "auto-upgrade", true, fmt.Sprintf("Auto-upgrade plugins' images to the current release %s", controllers.ImageMinVersion.String()))
```

mythi commented on pkg/controllers/reconciler.go:
```diff
@@ -159,6 +166,35 @@ func (r *reconciler) createObjects(ctx context.Context,
+		if parts := strings.SplitN(filepath.Base(*s), ":", 2); len(parts) == 2 && len(parts[0]) > 0 {
+			name, version := parts[0], parts[1]
+			if ver, err := versionutil.ParseSemantic(version); err == nil && ver.LessThan(ImageMinVersion) {
+				*s = "intel/" + name + ":" + ImageMinVersion.String()
```
we cannot change the repo org field. `intel/` is the default but users may have their own.

ozhuraki commented:
@rojkov @mythi @bartosh
Changes:
- Keep the [registry-url]/[namespace]/[image] intact, i.e. upgrade only the version

ozhuraki commented:
@rojkov @mythi @bartosh
Changes, please take a look:
- Updated documentation
- Dropped command line control
- All operator-managed plugins (deployed and new) are upgraded
- Updated envtests (daemonsets are injected)

mythi commented:
@ozhuraki thanks! looks OK at quick glance. I'll give it a try and take a closer review. can we fix that `fpgadeviceplugin` ds name in the YAML?

mythi commented on pkg/controllers/sgx/controller.go:
```diff
@@ -142,6 +147,11 @@ func (c *controller) UpdateDaemonSet(rawObj client.Object, ds *apps.DaemonSet) (
 		updated = true
 	}
 
+	if ds.Spec.Template.Spec.InitContainers[0].Image != dp.Spec.InitImage {
+		ds.Spec.Template.Spec.InitContainers[0].Image = dp.Spec.InitImage
+		updated = true
+	}
+
```
we need to check  `dp.Spec.InitImage == ""`.  GPU  controller logic can be followed here. let's have this fix as a separate commit

mythi commented on deployments/fpga_plugin/base/intel-fpga-plugin-daemonset.yaml:
```diff
@@ -1,7 +1,7 @@
 kind: DaemonSet
 metadata:
-  name: fpgadeviceplugin
+  name: intel-fpga-plugin
```
let's split these `intel-fpga-plugin` YAML changes to a separate commit so that the `suite_test.go` patch goes away.

mythi reviewed (commented):
@ozhuraki Looks good, I tested with QAT and it works. Two review comments about splitting changes to separate commits and one about the SGX controller fix. Thanks a lot!

ozhuraki commented:
@mythi
Thanks for the review!
Changes:
- Split and updated the SGX fix, re-split FGPA fixes and moved them upfront

rojkov reviewed (approved):
Thanks! LGTM

bart0sh reviewed (approved):
/lgtm
